Add warnings/errors for unreasonable DHW temperatures for WaterUse:Equipment

matthew-larson opened the pull request:
Pull request overview
---------------------
 - Fixes #7133 and #8806 
 - This pull request adds warnings for unreasonable DHW temperatures for the target temperature, hot water temperature, and cold water temperature.  A general warning will be included if it only occurs once, but a recurring warning will be generated if it occurs multiple times.  This will help prevent unreasonable water temperatures by warning the user if the hot water temperature is below the cold water temperature, and if the target temperature is below the cold water temperature or above the hot water temperature, making the target temperature unachievable and the energy consumption inaccurate.
 - Created a new unit test file for WaterUse.unit.cc that has a unit test that checks these three warnings.


### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally

matthew-larson committed:
Add warning/error if DHW hot water temperature is less than the cold water temperature.

matthew-larson committed:
Add unit test for hot/cold water temperature comparison.

matthew-larson committed:
Add warning for DHW water target temperature less than cold water temperature.

matthew-larson committed:
Add warning for DHW target temperature greater than hot water temperature.

matthew-larson committed:
Minor updates and new unit test for high DHW target temperature.

matthew-larson committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into add-waterusetemp-error

matthew-larson committed:
Fix clang formatting.

matthew-larson committed:
Fix clang formatting.

matthew-larson committed:
New unit test for low DHW target temperature.

matthew-larson committed:
Update description.

matthew-larson committed:
Add EPSILON constant to allow for near-zero comparison values.

matthew-larson committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into add-waterusetemp-error

matthew-larson commented:
All diffs are due to situations where the target water temperature is unachievable due to the hot water temperature, or the hot water temperature is below the cold water temperature.  Not sure if these should be investigated in the models to eliminate these warnings.

Nigusse commented on src/EnergyPlus/WaterUse.cc:
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
+                        this->HWTempErrIndex,
+                        this->HotTemp,
+                        this->HotTemp);
+                }
```
Catches when *Cold* water temp is > available *Hot* water temp and throws one time and recurring warning error message.

Nigusse commented on src/EnergyPlus/WaterUse.cc:
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
+                        this->TargetTempErrIndex,
+                        this->HotTemp,
+                        this->ColdTemp);
+                }
```
Catches when *Target* water temp is > available *Hot* water temp and throws one time and recurring warning error message.

Nigusse commented on src/EnergyPlus/WaterUse.cc:
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
+                        this->TargetTempErrIndex,
+                        this->HotTemp,
+                        this->ColdTemp);
+                }
```
Catches when *Target* water temp is < available *Cold* water temp and throws one time and recurring warning error message.

Nigusse commented on tst/EnergyPlus/unit/WaterUse.unit.cc:
```diff
@@ -0,0 +1,494 @@
+    });
+
+    EXPECT_TRUE(compare_err_stream(error_string3, true));
+}
```
This unit test demonstrates the three conditions where throwing warning error message is warranted. Update the branch, built and ran the unit test locally. Also ran the unit test suite locally and all passed.

Nigusse commented:
Now there are 35 example files that throw warning error message because they are caught by the fix. But these are expected error messages unless we want to modify the example files to eliminate the warning error. @Myoldmopar

Nigusse commented:
If the warning error message is a concern another alternative is to not report the warning error message unless "DisplayExtraWarnings" variable is set to true.

mjwitte commented:
@matthew-larson @Nigusse Ultimately, we want to fix these so the example files deliver the requested target temp without warnings. I would not hide these under "DisplayExtraWarnings". That could happen here or as a sperate defect/PR.

These are are all good warnings that indicate the target temp won't be met, and why. Do we have a warning that checks the supply temp vs target temp directly and warns that it didn't deliver? Or would that even be useful? If all of these checks pass, are there other things that could prevent the target temp from being met?

matthew-larson commented:
@mjwitte @Nigusse I ended up not doing a check for supply temp vs target temp as this would involve whole different part of the code and I thought would probably be better as a separate issue/pull request.  This could probably go along with DHW heating capacity checks which could be the other reason that the target temp wouldn't be reached.

I can go ahead and create another issue/pull request to fix the example files so we can push this one through sooner than later without me holding it up.

Nigusse commented:
@mjwitte @matthew-larson  I suspect sometimes the available capacity and demand mismatch may lead to hot water temperature to drop below the target temperature. But these needs to investigate case by case. I think removing the error message from the example files should be handled in a separate issue.

Nigusse commented:
@mjwitte I was making comment but ended up closing the issue. That was inadvertent.

Nigusse commented:
I looked at a couple of example files trying to understand the source and causes of the warning messages. For instance, the example file **ReportHeatEmission_RefFSRestaurant.idf** has a total of 41 warning massages and these messages occur during warmup period only. There are also two other observations I made:
- the warning was issued at Simulation time=07/21 00:00 - 00:15
- when I switched the order of the SizingPeriod:DesignDay (summer first) the warning error message occurred at Simulation time=01/21 00:00 - 00:15.  
- examining the hot water temperature in the tank during these hours indicate the hot water temperature was ~59C, way higher than the target temperature 49C. 
- This implies there is initialization problem.

@matthew-larson please add timestamp reporting function ` ShowContinueErrorTimeStamp(state, "") ` where you added the warning error message code (three locations). This will help us determine if the problem I am seeing is the same across all the example files with error file diffs and it is helpful to have timestamp for future use.

Nigusse commented on src/EnergyPlus/WaterUse.cc:
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
+                // print error for variables of hot water temperature
+                ++this->HWTempErrorCount;
+                if (this->HWTempErrorCount < 2) {
+                    ShowWarningError(state, "CalcEquipmentFlowRates: Hot water temperature is less than the cold water temperature");
```
Add this `ShowContinueErrorTimeStamp(state, "");` here

Nigusse commented on src/EnergyPlus/WaterUse.cc:
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
+                // print error for variables of target water temperature
+                ++this->TargetTempErrorCount;
+                if (this->TargetTempErrorCount < 2) {
+                    ShowWarningError(state, "CalcEquipmentFlowRates: Target water temperature is greater than the hot water temperature");
```
Add this `ShowContinueErrorTimeStamp(state, "");` here

Nigusse commented on src/EnergyPlus/WaterUse.cc:
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
+                // print error for variables of target water temperature
+                ++this->TargetTempErrorCount;
+                if (this->TargetTempErrorCount < 2) {
+                    ShowWarningError(state, "CalcEquipmentFlowRates: Target water temperature is less than the cold water temperature");
```
Add this `ShowContinueErrorTimeStamp(state, "");` here

matthew-larson committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into add-waterusetemp-error

matthew-larson committed:
Added timestamp to warning error and updated unit test accordingly.

matthew-larson commented:
Thanks @Nigusse for your review.  I went ahead and added the `ShowContinueErrorTimeStamp(state, "");` lines as suggested, updated the unit tests accordingly, and merged in the latest develop branch.

Myoldmopar commented:
I'll pull and test this locally.

Myoldmopar commented:
Builds and passes locally with develop pulled in.  Regressions locally are exactly what it was showing the last time Decent ran.  Merging, thanks @matthew-larson and @Nigusse!

matthew-larson commented:
Thanks @Myoldmopar!
